## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv',low_memory=False, index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
print(contrib.shape)

(1125659, 18)


- **1a answer here:** 
- **Yes this matches the expectation.**

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
print(contrib.columns.values.tolist())

['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city', 'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation', 'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd', 'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp']


- **1b answer here:** 
- **Yes, All columns are included. **

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
print(contrib.head(5))

     cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st contbr_zip            contbr_employer  \
0       LARKSPUR        CA  949391913                        NaN   
1        CAMBRIA        CA  934284638                        NaN   
2        FONTANA        CA  923371507                        NaN   
3      CAMARILLO        CA  930111214  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA  902784310   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1 

- **1c answer here:** 
- **Candidate Id, Name and State have values as per their column definition. They hold unique Id for candidate, their name and state from which they are contesting elections. **

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
print(contrib.election_tp.unique())

['P2016' 'G2016' nan 'P2020']


- **1d answer here:** 
- **Distinct values in election_tp column indicates whether contribution was made for general election or primary election. Next four digits indicate the year of contribution.**

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
print(contrib.dtypes)

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object


- **1e answer here:** 
- **contbr_zip = string, contb_receipt_amt = float, contb_receipt_dt = datetime**

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
print(contrib.isnull().sum().sort_values(ascending=0))

receipt_desc         1110614
memo_cd               981391
memo_text             624511
contbr_employer       157902
contbr_occupation      10399
election_tp             1425
contbr_zip                95
contbr_city               26
tran_id                    0
file_num                   0
form_tp                    0
cmte_id                    0
contb_receipt_dt           0
cand_id                    0
contbr_st                  0
contbr_nm                  0
cand_nm                    0
contb_receipt_amt          0
dtype: int64


- **1f answer here:** 
- **receipt_desc, memo_cd and memo_text are top three null value columns. I would consider receipt_desc column based on number of null values and their minimal significance on core data. contbr_employer and contbr_occupation are having null values as well but they are less than 10% of the total data and can be used to promote corporate level initiatives or occupation level initiatives in future. memo_cd and memo_text can be refferred to find transaction source or previous step of a particular row.**

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE

print(contrib.groupby(['cand_id','cand_nm']).agg({'cand_id': ['nunique', 'count'],'cand_nm': ['nunique', 'count']}))

                                    cand_id         cand_nm        
                                    nunique   count nunique   count
cand_id   cand_nm                                                  
P00003392 Clinton, Hillary Rodham         1  547211       1  547211
P20002671 Johnson, Gary                   1    1591       1    1591
P20002721 Santorum, Richard J.            1      88       1      88
P20003281 Perry, James R. (Rick)          1     116       1     116
P20003984 Stein, Jill                     1    1907       1    1907
P40003576 Paul, Rand                      1    4268       1    4268
P60003670 Kasich, John R.                 1    3005       1    3005
P60005915 Carson, Benjamin S.             1   27362       1   27362
P60006046 Walker, Scott                   1     737       1     737
P60006111 Cruz, Rafael Edward 'Ted'       1   57820       1   57820
P60006723 Rubio, Marco                    1   14092       1   14092
P60007168 Sanders, Bernard                1  407

- **1g answer here:** 
- **Above output represents values grouped by candidate id. Sanders, Bernard is uniquely reprented by P60007168 id throughout the data. Count on the right side confirms there are no issues with any other candidate id and candidate name.**

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
print(contrib['contbr_st'].value_counts())

CA    1125659
Name: contbr_st, dtype: int64


- **1h answer here:** 
- **Based on above output there are no records outside the state of CA based on contbr_st.**

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
print(contrib.tran_id.duplicated().value_counts())
print(contrib[contrib.duplicated(['tran_id'])==True].sort_values('tran_id'))

False    1122205
True        3454
Name: tran_id, dtype: int64
           cmte_id    cand_id                    cand_nm  \
43573    C00575449  P40003576                 Paul, Rand   
42835    C00575449  P40003576                 Paul, Rand   
44679    C00575449  P40003576                 Paul, Rand   
42820    C00575449  P40003576                 Paul, Rand   
44680    C00575449  P40003576                 Paul, Rand   
...            ...        ...                        ...   
592806   C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1024607  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
341445   C00573519  P60005915        Carson, Benjamin S.   
37477    C00573519  P60005915        Carson, Benjamin S.   
37470    C00573519  P60005915        Carson, Benjamin S.   

                    contbr_nm   contbr_city contbr_st contbr_zip  \
43573    PARKS, PATRICIA MRS.       HUGHSON        CA      95326   
42835    PARKS, PATRICIA MRS.       HUGHSON        CA      95326   
44679        

- **1i answer here:**
- **Above output indicates there are 3454 duplicate values for tran_id column hence this column alone may not be used as primary key. Some of the rows are change from primary to general and vice-versa and some rows are duplicate rows ** 

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE
print(contrib.fillna(0).query('contb_receipt_amt < 0'))

           cmte_id    cand_id                    cand_nm  \
19       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
23       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
81       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
190      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
213      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125008  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125317  C00580399  P60008521   Christie, Christopher J.   
1125427  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125446  C00573519  P60005915        Carson, Benjamin S.   
1125582  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   

                    contbr_nm             contbr_city contbr_st contbr_zip  \
19           JOLLIFF, RICHARD                   CHICO        CA  959289507   
23           JOLLIFF, RICHARD                   CHICO        CA  959289507   
81           JOLLIFF, RICHARD                

- **1j answer here:**
- **As per output 11896 rows are havving neagative amounts. These rows are redesignation to general or are refund amounts. These amounts do match with the expectation.**

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE
start_date = '2014-01-01'
end_date   = '2016-06-07'
print(contrib.query('contb_receipt_dt < @start_date and contb_receipt_dt > @end_date'))

Empty DataFrame
Columns: [cmte_id, cand_id, cand_nm, contbr_nm, contbr_city, contbr_st, contbr_zip, contbr_employer, contbr_occupation, contb_receipt_amt, contb_receipt_dt, receipt_desc, memo_cd, memo_text, form_tp, file_num, tran_id, election_tp]
Index: []


- **1k answer here:**
- **No date is outside 1 Jan 2014 to 7 June 2016.**

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:**
- **Yes we have correct number of columns and rows.**

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:**
- **Yes records contain data for the questions we need to address.**

**1l.3** What columns are important?

- **1l.3 answer here:** 
- **cmte_id, cand_id, cand_nm, contbr_nm, contbr_city, contbr_st, contbr_zip, contb_receipt_amt, contb_receipt_dt,  memo_cd, memo_text, form_tp, tran_id, election_tp columns are important.**

**1l.4** What columns can be dropped?

- **1l.4 answer here:** 
- **receipt_desc column can be dropped**

**1l.5** What are the data problems?

- **1l.5 answer here:**
- **Some of the observations are: memo_text and receipt_desc have redundant data. Zip codes are not five digit or uniform. There are duplicate rows in the file considering few rows have exactly same data for a transaction id.**

**1l.6** List any assumptions so far:

- **1l.6 answer here:**
- **Benard Sandres  is not same as Bernie Sanders.**

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [13]:
# 2a YOUR CODE HERE
df5 = contrib[contrib.election_tp.str[0] != 'P']
print(df5.shape)

(315171, 18)


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [101]:
# 2b YOUR CODE HERE
list_values = ["Bernie Sanders"]
df6 = df5[~df5['cand_nm'].isin(list_values)]
print(df6.shape)

(315171, 18)


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [123]:
# 2c YOUR CODE HERE
df8 = df6
df8['contbr_zip_5'] = df8['contbr_zip'].str[:5].fillna(0).astype(int) #fillna is used to replace NAN values with zero so that new column can be casted as int.
df9 = df8.query('contbr_zip_5 >= 90001 and contbr_zip_5 <= 96162' ) #This range is for CA Zipcodes
print(df9.shape)


(314976, 19)


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [118]:
# 2d YOUR CODE HERE
df10 = df9.query('contb_receipt_amt >= 0')
print(df10.shape)

(311624, 19)


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [119]:
# 2e YOUR CODE HERE
#receipt_desc         1110614
df11 = df10.drop(['receipt_desc'], axis=1)
print(df11.shape)

(311624, 18)


- **2e answer here:**
- **receipt_desc is dropped as it has maximum null values and memo_text column has same data as receipt_desc for most of the non-null rows.

**2f.** List any assumptions that you made up to this point:

- **2f answer here:**
- **Sandres Benard is not same as Bernie Sanders. Duplicate transaction id rows are still there**

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [120]:
# 3a YOUR CODE HERE
df12=df11.groupby('contbr_zip_5')['contb_receipt_amt'].count().sort_values(ascending=False) # Count of contribution, hence count of amount is taken. Other column can be taken as well.
df13=df11.groupby('contbr_zip_5')['contb_receipt_amt'].sum().sort_values(ascending=False)
print(df12) #Can print top five rows using head to keep it clean but wanted to see larger output.
print(df13) #Can print top five rows using head to keep it clean but wanted to see larger output.


contbr_zip_5
94114    3387
94110    3027
94611    2602
90046    2178
90069    2033
         ... 
92605       1
92581       1
95526       1
95528       1
95487       1
Name: contb_receipt_amt, Length: 1928, dtype: int64
contbr_zip_5
94114   702,180.21
94301   638,905.06
90049   624,604.93
94110   539,026.32
90210   534,369.86
           ...    
93242         8.00
95412         8.00
95234         4.00
96035         4.00
95528         0.80
Name: contb_receipt_amt, Length: 1928, dtype: float64


- **3a answer here:** 
- **94114 Zip code has highest count of contributions. 94114 Zip code has most dollar amount of contribution considering data filters which were asked to be applied.**

**3b.** What day(s) of the month do most people donate? (10 points)

In [121]:
# 3b YOUR CODE HERE

df23=df11.groupby(['contb_receipt_dt'])['contb_receipt_amt'].sum().sort_values(ascending=False)
print(df23)


contb_receipt_dt
2016-09-30   1,188,839.67
2016-09-26   1,076,952.45
2016-07-29     995,646.64
2016-09-27     908,279.98
2016-09-28     839,150.94
2016-08-31     810,353.64
2016-07-28     809,420.71
2016-10-19     807,885.15
2016-09-29     801,298.64
2016-09-16     627,262.78
2016-10-18     626,030.91
2016-08-01     614,006.92
2016-09-20     603,899.33
2016-08-09     603,064.79
2016-09-15     595,681.02
2016-09-21     583,877.33
2016-08-03     579,306.82
2016-09-23     578,284.75
2016-09-14     577,495.39
2016-07-31     571,728.34
2016-08-05     560,565.76
2016-10-09     547,271.12
2016-09-19     546,551.96
2016-08-02     544,936.51
2016-08-08     539,849.54
2016-10-10     535,070.12
2016-08-15     532,631.31
2016-09-07     517,680.27
2016-08-10     511,511.84
2016-08-12     509,880.06
2016-08-04     506,199.30
2016-09-12     499,940.83
2016-08-17     499,165.78
2016-07-30     490,594.83
2016-08-22     486,196.00
2016-08-16     484,970.45
2016-08-30     484,917.41
2016-10-17     483,36

- **3b answer here:** 
- **Above output indicates that top five contributions were made on the last four days working days of the month. These days also could be payroll processing days for people who are on bi-weekly payroll. 07/29 and 09/30 are last working day of the month as well as payroll day for most of the pay cycles. 09/26 is monday and is payroll processing days for companies having monday pay cycle. This has resulted in maximum contribution on these days.

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k